File: enum-jpa-validation-app/src/main/java/com/example/enumapp/web/validation/LocalDateString.java
package com.example.enumapp.web.validation;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * null/blank 는 통과. 값이 있으면 yyyy-MM-dd 이고 실제 LocalDate 로 파싱 가능해야 함.
 */
@Documented
@Target({ElementType.FIELD, ElementType.PARAMETER})
@Retention(RetentionPolicy.RUNTIME)
@Constraint(validatedBy = LocalDateStringValidator.class)
public @interface LocalDateString {

    String message() default "must match LocalDate pattern yyyy-MM-dd";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};
}

File: backend/enum-jpa-querydsl-kotlin-app/src/main/kotlin/com/example/enumapp/web/validation/LocalDateString.kt
package com.example.enumapp.web.validation

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.annotation.AnnotationRetention.RUNTIME
import kotlin.annotation.AnnotationTarget.FIELD
import kotlin.annotation.AnnotationTarget.VALUE_PARAMETER
import kotlin.reflect.KClass

/**
 * null/blank 는 통과. 값이 있으면 yyyy-MM-dd 이고 실제 LocalDate 로 파싱 가능해야 함.
 */
@MustBeDocumented
@Target(FIELD, VALUE_PARAMETER)
@Retention(RUNTIME)
@Constraint(validatedBy = [LocalDateStringValidator::class])
annotation class LocalDateString(
    val message: String = "must match LocalDate pattern yyyy-MM-dd",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = [],
)
